### ISSUE TITLE
breakpoint doesn't work in scripts

### ISSUE THREAD

**Issue by InspectorMustache** [fish-shell/fish-shell#4823]
I might be missing something but I'm having trouble getting breakpoint to work when I use it in scripts instead of functions.

This works fine:
```
~> function blah
       breakpoint
       echo "after breakpoint"
   end
~> blah
BP blah:3 >  (Ctrl-D)
after breakpoint
```

However there's no prompt when I do the same from within a script file:
```
~> echo '#!/bin/fish
   breakpoint
   echo "after breakpoint"' > blah.fish
~> chmod +x blah.fish
~> ./blah.fish 
after breakpoint
```

I'm using fish 2.7.1 but I also tried the latest git version and had no luck as well.

**Comment by faho** [comment]
It's currently explicitly disabled when not running interactively.

I'm not sure how we want to handle this. Should it always be enabled? Only when stdin is a tty? Only when a special option has been given or variable has been set?

I'd love to know how people use `breakpoint`, or would use it if it were changed a bit. It's a cool idea that I believe could do with some improvement.

**Comment by InspectorMustache** [comment]
I don't know what the implications of enabling it when not running interactively are but I think debugging scripts is one of the main use cases for a command like breakpoint. The implementation of breakpoint in fish is a really nice feature and it would be quite a waste if scripts couldn't benefit from it.
In regards to the command itself, I think running standard shell commands and looking up variables is already really useful. Step-through debugging would be amazing of course, but much harder to implement, I would guess.

**Comment by jay-tuckey** [comment]
This would be very useful for me, to help with debugging scripts as I write them. Perhaps the `breakpoint` command could have an `--always` flag that will start debugging even when not interactive?

**Comment by faho** [comment]
Okay. How about this:

Any given breakpoint is enabled iff `$fish_breakpoints` contains the name of the function that it is currently in. So if you want to break in the function `dostuff`, you `set -a fish_breakpoints dostuff`. If there is more than one breakpoint, it breaks for both. If there is another function with the same name defined later, it also breaks there.

I would probably remove the interactivity check entirely, which would mean that you'd have to set the variable to arm the breakpoint. It might be possible to add a special value to enable *all* breakpoints, though if you can select them that might lead to people using more breakpoints, which would make that unattractive.

Alternatively, we could add another builtin to select breakpoints to enable, or we could add options to `breakpoint` to do that (like `breakpoint --name foo` to name a breakpoint and `breakpoint enable foo` later), or we could just always enable them and require users to use `if something; breakpoint`. These options all seem either too complicated (I don't think selecting only certain breakpoints in a function is worth it), or have drawbacks (`if` makes a new block, so it has issues with local variables).

**Comment by jay-tuckey** [comment]
How would this work for script files? Would you be able to have script file `update.fish` and in it use `set -a fish_breakpoints update.fish; breakpoint` or something similar?

**Comment by zanchey** [comment]
What about a switch for the fish executable itself? `fish --with-breakpoints`?

**Comment by faho** [comment]
>What about a switch for the fish executable itself? fish --with-breakpoints?

The problem with enabling all breakpoints globally is that you might want to leave breakpoints in. E.g. if I leave a breakpoint in my prompt, starting fish with that enabled means I have to exit that breakpoint every time.

Which is why I think we want to be able to select breakpoints.

Also requiring a restart is a limitation in itself - you might want to enable a breakpoint and require some environmental setup that a restart would undo, so you'd have to redo it.

**Comment by jay-tuckey** [comment]
> The problem with enabling all breakpoints globally is that you might want to leave breakpoints in.

Perhaps this could be better addressed by having people leaving breakpoints in having them be conditional. For example, if I have a script and want to leave a permanent breakpoint, I do something like:
```
#!/usr/bin/env fish

# do some argparsing

if set --query debugmode
    breakpoint
end
```

This approach would keep the breakpoint logic really simple, and keep the logic of run/don't run the breakpoint right next to the breakpoint itself.

My feeling is that breakpoints are a very temporary thing that you remove at the end, but I may have different use-cases from others. What are your thoughts on this approach?

**Comment by cben** [comment]
>  So if you want to break in the function dostuff, you `set -a fish_breakpoints dostuff`.

I think that's too much work and hard to discover.  Many users would just conclude, "oh well it didn't work" and give up :-(  
The simplest & least surprising experience for debugging is: add `breakpoint` anywhere => it stops there :stop_sign: 

The 2nd use case, which is quite common (at least from experience in other languages), is: the code is running but a breakpoint you'd inserted is too "spammy" — you already understood what happens there, but it keeps stopping there again and again.  For example if it's in a loop.  So you *couldn't have planned for it* but now want to disable it.
So yeah this would be nice but I propose should be opt-out not opt-in.

The 3rd use case of leaving "dormant" breakpoint in scripts/functions for a long time is I think "advanced usage", can be achieved in several ways but usability is less important.

**Comment by InspectorMustache** [comment]
I agree. For me, a breakpoint should just cause a stop whenever it's encountered. Everything else seems a bit convoluted to me. If a more complex logic is desired (conditionals etc.) then I think it's okay if that is reflected by more complex code.

**Comment by zanchey** [comment]
@ridiculousfish, any thoughts on this?

**Comment by faho** [comment]
Okay, how about this:

- `breakpoint` takes an optional name. If that name is in $fish_breakpoints, it breaks. (this should probably be `--name=` so we don't remove the possibility of future subcommands here)
- If no name is given, it breaks regardless
- $fish_breakpoints is defined to be a path variable, so calling `fish_breakpoint=foo:bar fish` enables both foo and bar
- breakpoint breaks when stdin is a tty, but the shell does not have to be interactive. There's no use in piping scripts to breakpoint.

That way we could add easy-access named breakpoints that can be selected, and simple breakpoints just break, and if more complicated conditions are needed you can just check them - `status is-interactive; or breakpoint`.

**Comment by marcusatiliusregulus** [comment]
sounds good, this way it works as expected and configuration of debugging options via fish_* variables is known from fish_trace

**Comment by thomasbbrunner** [comment]
The documentation for `breakpoint` in the [website](https://fishshell.com/docs/current/language.html#debugging) is misleading. It states that: "To start a debug session simply run the builtin command breakpoint at the point in a function *or script* where you wish to gain control."

Also, even if I run the script with the `--interactive` argument, it still does not stop at the breakpoint. Is there a way to force fish to stop in a breakpoint?

**Comment by werifu** [comment]
> Currently, breakpoints are disabled unless it's actually in an interactive shell. That applies even to shells started with `--interactive`. That is in fact what this issue is about.

Does fish have a plan to support paused at breakpoints even then not in interactive mode? Most of people in this issue may support that...

**Comment by ridiculousfish** [comment]
Nobody's working on this particular issue but I think we'd all like a fix.
